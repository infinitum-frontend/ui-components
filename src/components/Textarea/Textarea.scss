@import '../Input/common.scss';
.inf-textarea {
  box-sizing: border-box;
  outline: none;
  cursor: text;
  display: block;
  width: 100%;
  border: 1px solid var(--inf-color-border-default);
  border-radius: var(--inf-border-radius-small);
  transition: border-color 0.2s;
  font-weight: var(--inf-font-weight-normal);
  font-size: var(--inf-text-size-small);
  line-height: 18px;
  padding: 7px 12px;
  background-color: inherit;
  overflow: hidden;
  resize: none;

  &:hover {
    border-color: var(--inf-color-border-hover);

    &::placeholder {
      color: var(--inf-color-text-control);
    }
  }

  &:focus {
    border-color: var(--inf-color-border-focused);
  }

  &:disabled {
    background-color: var(--inf-input-color-background-disabled);
    pointer-events: none;
    border-color: transparent;
    color: var(--inf-color-text-quaternary);

    &::placeholder {
      color: var(--inf-color-text-quaternary);
    }
  }

  &--status-error {
    border-color: var(--inf-color-primary);
  }

  &--block {
    width: 100%;
  }

  &--filled {
    border-color: var(--inf-color-border-focused);
  }

  &--resize-both {
    resize: both;
  }

  &--resize-horizontal {
    resize: horizontal;
  }

  &--resize-vertical {
    resize: vertical;
  }
}
