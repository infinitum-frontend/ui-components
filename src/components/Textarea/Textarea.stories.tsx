import { StoryFn, Meta } from '@storybook/react'
import { Textarea } from './index'

const meta: Meta<typeof Textarea> = {
  title: 'Textarea',
  component: Textarea,
  args: {
    disabled: false
  },
  argTypes: {
    cols: {
      type: 'number'
    },
    rows: {
      type: 'number'
    }
  }
}

export default meta

const Template: StoryFn<typeof Textarea> = (args) => {
  return <Textarea {...args} />
}

export const Playground = Template.bind({})

export const Disabled = Template.bind({})
Disabled.args = { disabled: true }

export const Block = Template.bind({})
Block.args = { block: true }

export const Resizable = Template.bind({})
Resizable.args = { resize: 'both' }
