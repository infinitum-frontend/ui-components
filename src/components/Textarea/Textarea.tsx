// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, {
  FormEvent,
  FormEventHandler,
  ReactElement,
  TextareaHTMLAttributes,
  useContext
} from 'react'
import cn from 'classnames'
import './Textarea.scss'
import { TextFieldClasses } from '~/src/utils/textFieldClasses'
import FormGroupContext from 'Components/Form/context/group'
import FormContext from 'Components/Form/context/form'
import useFormControlHandlers from 'Components/Form/hooks/useFormControlHandlers'

export interface TextareaProps
  extends Omit<
    TextareaHTMLAttributes<HTMLTextAreaElement>,
    'onInput' | 'onChange'
  > {
  /** Значение */
  value?: string
  onInput?: (value: string, event: FormEvent<HTMLTextAreaElement>) => void
  onChange?: (value: string, event: FormEvent<HTMLTextAreaElement>) => void
  /** Состояние (ошибка) */
  status?: 'error'
  /**
   * Состояние недоступности
   * @default false
   */
  disabled?: boolean
  /**
   * Плейсхолдер
   * @default 'Введите значение'
   */
  placeholder?: string
  /**
   * Возможные варианты ресайза
   * @default 'none'
   */
  resize?: 'none' | 'vertical' | 'horizontal' | 'both'
  /**
   * Занимает ли контент всю ширину
   * @default false
   */
  block?: boolean
}

/** Компонент ввода для написания объемного текста */
const Textarea = ({
  resize = 'none',
  block = false,
  disabled: disabledProp = false,
  placeholder = 'Введите значение',
  defaultValue,
  value,
  className,
  status,
  onInput,
  onChange,
  id,
  required = false,
  'aria-required': ariaRequired,
  'aria-invalid': ariaInvalid,
  ...props
}: TextareaProps): ReactElement => {
  const formData = useContext(FormContext)
  const formGroupData = useContext(FormGroupContext)
  const { onControlChange, onControlInvalid } = useFormControlHandlers()

  const disabled = disabledProp || formData?.disabled

  const handleChange: FormEventHandler<HTMLTextAreaElement> = (e) => {
    onControlChange(e)

    const domValue = (e.target as HTMLTextAreaElement).value
    onChange?.(domValue, e)
    onInput?.(domValue, e)
  }

  const classNames = cn(
    'inf-textarea',
    TextFieldClasses.main,
    TextFieldClasses.borderRadius.regular,
    className,
    {
      'inf-textarea--block': block,
      [TextFieldClasses.status[status as 'error']]: status,
      [`inf-textarea--resize-${resize as string}`]: resize
    }
  )

  return (
    <textarea
      placeholder={placeholder}
      className={classNames}
      defaultValue={defaultValue}
      value={value}
      id={id || formGroupData?.id}
      onChange={handleChange}
      onInvalid={onControlInvalid}
      disabled={disabled}
      required={formGroupData?.required || required}
      aria-required={formGroupData?.required || ariaRequired}
      aria-invalid={formGroupData?.invalid || ariaInvalid}
      {...props}
    />
  )
}

export default Textarea
