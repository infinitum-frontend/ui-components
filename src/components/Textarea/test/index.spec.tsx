import { it, describe, expect } from 'vitest'
import { render } from '@testing-library/react'
import { Textarea } from '../index'
import { TextareaProps } from '../Textarea'

const renderComponent = (
  args?: TextareaProps
): Record<'el' | 'container', HTMLElement> => {
  const { container } = render(<Textarea {...args} />)

  const el = container.firstChild as HTMLElement

  return { el, container }
}

describe('Textarea', () => {
  it('should render', () => {
    const { el } = renderComponent()
    expect(el).toBeDefined()
    expect(el.className).toContain('inf-textarea')
  })

  it('should match snapshot', () => {
    const { el } = renderComponent()
    expect(el).toMatchSnapshot()
  })

  it('should support error status', () => {
    const { el } = renderComponent({ status: 'error' })
    expect(el.className).toContain('inf-textarea--status-error')
    expect(el).toHaveStyle('border-color: var(--inf-color-primary);')
  })

  it('should support full width', () => {
    const { el } = renderComponent({ block: true })
    expect(el.className).toContain('inf-textarea--block')
    expect(el).toHaveStyle('width: 100%')
  })

  it('should support placeholder', () => {})
})
