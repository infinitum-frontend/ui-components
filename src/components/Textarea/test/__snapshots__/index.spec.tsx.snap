// Vitest Snapshot v1

exports[`Textarea > should match snapshot 1`] = `
<textarea
  class="inf-textarea inf-input-common inf-textarea--resize-none"
  placeholder="Введите значение"
/>
`;
